middleware: add sanitization hook

mrd0ll4r opened the pull request:
This adds

1. Benchmarks for hooks. This helps estimate the overhead of our middleware-based approach and to make possible optimizations (i.e. reuse the `AnnounceResponse` struct).
2. An `AddressFamily` field for `Peer`s.
3. A hook that checks a Peer's IP and sets the above field, and also checks numWant.

Also, can a native speaker please tell me whether to name this `sanitationHook` `sanitizationHook` or something else entirely? That'd be great :)

This is about #254 mostly


Results of three benchmark runs on my laptop: (`none` has no middleware at all, just the `AnnounceResponse` alloc, _k_`nop` has _k_ no-op hooks chained, _k_`sanitation` has _k_ sanitation hooks chained)
```
name                         time/op
HookOverhead/none-v4         145ns ± 3%
HookOverhead/none-v6         146ns ± 1%
HookOverhead/1nop-v4         154ns ± 2%
HookOverhead/1nop-v6         155ns ± 0%
HookOverhead/2nop-v4         159ns ± 2%
HookOverhead/2nop-v6         157ns ± 1%
HookOverhead/3nop-v4         163ns ± 4%
HookOverhead/3nop-v6         166ns ± 4%
HookOverhead/1sanitation-v4  168ns ± 0%
HookOverhead/1sanitation-v6  164ns ± 0%
HookOverhead/2sanitation-v4  174ns ± 1%
HookOverhead/2sanitation-v6  180ns ± 2%
HookOverhead/3sanitation-v4  188ns ± 4%
HookOverhead/3sanitation-v6  196ns ± 1%
```
For whomever is interested: I attached the raw benchmark results as [bench.txt](https://github.com/chihaya/chihaya/files/617357/bench.txt).

The PR is WIP. In case we move forward with this:

- [x] Replace usages of `len(Peer.IP)`
- [x] Remove IP checks in the frontend
- [x] Remove `numWant` checks in the storage
- [x] (optional) Add a `defaultNumWant` config field

jzelinskie commented on bittorrent/bittorrent.go:
```diff
@@ -110,12 +110,22 @@ type Scrape struct {
+
+// AddressFamily constants.
+const (
+	IPv4 = AddressFamily(iota)
```
That's interesting syntax.
I've always written enums `IPv4 AddressFamily = iota`
I'm not sure which is more idiomatic.

jzelinskie commented:
I think we should make a new type that is:

```go
type IP struct {
  net.IP
  AddressFamily
}
```

mrd0ll4r commented:
Updated, ptal.

Also fixed some bugs about determining whether scrapes are IPv6. Also cleaned up some unused constants in the UDP frontend.

This is a big one, hope I didn't forget something somewhere.

jzelinskie commented on middleware/hooks.go:
```diff
@@ -123,13 +164,13 @@ func (h *responseHook) appendPeers(req *bittorrent.AnnounceRequest, resp *bittor
 		resp.IPv6Peers = peers
 	default:
-		panic("peer IP is not IPv4 or IPv6 length")
+		panic("peer IP is not IPv4 or IPv6")
```
>attempted to append peer that is neither IPv4 nor IPv6

jzelinskie commented on middleware/middleware.go:
```diff
@@ -16,6 +16,8 @@ import (
 type Config struct {
 	AnnounceInterval time.Duration `yaml:"announce_interval"`
+	MaxNumWant       uint32        `yaml:"max_numwant"`
+	DefaultNumWant   uint32        `yaml:"default_numwant"`
```
If `Response` became its own middleware, these values would probably belong to that.

mrd0ll4r replied:
We could put them there - that would however make it impossible for a middleware to change them (the response middleware is just gonna use the configured values).

jzelinskie commented on middleware/middleware.go:
```diff
@@ -26,7 +28,7 @@ func NewLogic(cfg Config, peerStore storage.PeerStore, preHooks, postHooks []Hoo
 		announceInterval: cfg.AnnounceInterval,
 		peerStore:        peerStore,
-		preHooks:         append(preHooks, &responseHook{store: peerStore}),
+		preHooks:         append([]Hook{&sanitizationHook{maxNumWant: cfg.MaxNumWant, defaultNumWant: cfg.DefaultNumWant}}, append(preHooks, &responseHook{store: peerStore})...),
```
This line is getting pretty ugly. It should probably be broken it out.

jzelinskie commented on storage/memory/peer_store.go:
```diff
@@ -96,20 +93,30 @@ func (s *peerStore) shardIndex(infoHash bittorrent.InfoHash, v6 bool) uint32 {
 
 func decodePeerKey(pk serializedPeer) bittorrent.Peer {
-	return bittorrent.Peer{
+	toReturn := bittorrent.Peer{
```
s/toReturn/peer

jzelinskie commented on storage/memory/peer_store.go:
```diff
@@ -235,7 +242,7 @@ func (s *peerStore) GraduateLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) e
 	pk := newPeerKey(p)
 
-	shard := s.shards[s.shardIndex(ih, len(p.IP) == net.IPv6len)]
+	shard := s.shards[s.shardIndex(ih, p.IP.AddressFamily == bittorrent.IPv6)]
```
i wonder if just passing the AF and switching on it is faster

mrd0ll4r commented on middleware/hooks.go:
```diff
@@ -70,6 +70,47 @@ func (h *swarmInteractionHook) HandleScrape(ctx context.Context, _ *bittorrent.S
+		req.NumWant = h.maxNumWant
+	}
+
+	if req.NumWant == 0 {
```
Well, we need a default somewhere, we could do it in the storage, but then again every storage would have to do it - we can do it just once in here :)

mrd0ll4r commented on middleware/hooks.go:
```diff
@@ -97,7 +138,7 @@ func (h *responseHook) HandleAnnounce(ctx context.Context, req *bittorrent.Annou
 
 	// Add the Scrape data to the response.
-	s := h.store.ScrapeSwarm(req.InfoHash, len(req.IP) == net.IPv6len)
+	s := h.store.ScrapeSwarm(req.InfoHash, req.IP.AddressFamily == bittorrent.IPv6)
```
Probably, good point.

mrd0ll4r commented:
This now also addresses #264 .
Also updated, ptal.
